internal/scan/scan.go

```diff
@@ -70,27 +73,129 @@ type EventApprovalForAll struct {
 	Approved bool
 }
 
+// EventNewERC721BridgelessMinting is the ERC721 event emitted when a new Bridgeless Minting contract is deployed
+type EventNewERC721BridgelessMinting struct {
+	NewContractAddress common.Address
+	BaseURI            string
+}
+
 // Scanner is responsible for scanning and retrieving the ERC721 events
 type Scanner interface {
+	ScanNewBridgelessMintingEvents(ctx context.Context, fromBlock, toBlock *big.Int) error
 	ScanEvents(ctx context.Context, fromBlock *big.Int, toBlock *big.Int) ([]Event, error)
 }
 
 type scanner struct {
-	client   EthClient
-	contract common.Address
+	client    EthClient
+	contracts []common.Address
+	storage   Storage
 }
 
 // NewScanner instantiates the default implementation for the Scanner interface
-func NewScanner(client EthClient, contract common.Address) Scanner {
-	return scanner{
-		client:   client,
-		contract: contract,
+func NewScanner(client EthClient, s Storage, contracts ...string) Scanner {
+	scan := scanner{
+		client:  client,
+		storage: s,
 	}
+	for _, c := range contracts {
+		scan.contracts = append(scan.contracts, common.HexToAddress(c))
+	}
+	return scan
+}
+
+// ScanEvents returns the ERC721 events between fromBlock and toBlock
+func (s scanner) ScanNewBridgelessMintingEvents(ctx context.Context, fromBlock, toBlock *big.Int) error {
+	triggerDiscovery, err := s.triggerDiscovery(ctx)
+	if err != nil {
+		return err
+	}
+	if !triggerDiscovery {
+		return nil
+	}
+	eventLogs, err := s.filterEventLogs(ctx, fromBlock, toBlock, s.contracts...)
+	if err != nil {
+		return fmt.Errorf("error filtering events: %w", err)
+	}
+	if len(eventLogs) == 0 {
+		slog.Debug("no events found for block range", "from_block", fromBlock.Int64(), "to_block", toBlock.Int64())
+		return nil
+	}
+
+	contractAbi, err := abi.JSON(strings.NewReader(ERC721BridgelessMinting.ERC721BridgelessMintingMetaData.ABI))
+	if err != nil {
+		return fmt.Errorf("error instantiating ABI: %w", err)
+	}
+	for i := range eventLogs {
+		slog.Info("scanning event", "block", eventLogs[i].BlockNumber, "txHash", eventLogs[i].TxHash)
+		if len(eventLogs[i].Topics) == 0 {
+			continue
+		}
+
+		switch eventLogs[i].Topics[0].Hex() {
+		case eventNewERC721BridgelessMintingSigHash:
+			newERC721BridgelessMinting, err := parseNewERC721BridgelessMinting(&eventLogs[i], &contractAbi)
+			if err != nil {
+				return err
+			}
+			slog.Info("received event", eventNewERC721BridgelessMintingName, newERC721BridgelessMinting)
+
+			c := ERC721BridgelessContract{
+				Address: newERC721BridgelessMinting.NewContractAddress,
+				Block:   eventLogs[i].BlockNumber,
+				BaseURI: newERC721BridgelessMinting.BaseURI,
+			}
+			if err := s.storage.Store(ctx, c); err != nil {
+				return err
+			}
+		default:
+			slog.Debug("no new bridgeless minting contracts found")
+		}
+	}
+
+	return nil
+}
+
+func (s scanner) triggerDiscovery(ctx context.Context) (bool, error) {
```

Comment:
this name is misleading since it is no triggering the discovery. I would suggest to give it a more accurate name